osquery/events/linux/http_event_publisher.cpp

```diff
@@ -0,0 +1,863 @@
+     }
+
+     if (pcap_compile(
+             handle_, &fp_, sFilters.c_str(), 0, PCAP_NETMASK_UNKNOWN) == -1) {
+       LOG(ERROR) << "Could not compile filter expression";
+       return Status(1, "Could not compile filter expression.");
+     }
+     
+     if (pcap_setfilter(handle_, &fp_) == -1) {
+         LOG(ERROR) << "Could not install filter";
+         return Status(1, "Could not install filter.");
+     }
+     
+     int rc = pcap_loop(handle_, -1, processPacket, (unsigned char*)this);
+     if (rc == -1) {
+         LOG(ERROR) << "Error running pcap loop";
+         return Status(1, "Error running pcap loop.");
+     } else {
+         LOG(INFO) << "Stopping pcap loop";
+         return Status(2, "Stopping pcap loop.");
+     }
+ }
+ 
+ void HTTPLookupEventPublisher::stop() {
+     if (handle_ != nullptr) {
+         TLOG << "Stopping HTTP publisher";
+         
+         pcap_breakloop(handle_);
+         pcap_freecode(&fp_);
+         pcap_close(handle_);
+         handle_ = nullptr;
+     }
+ }
+ 
+ static inline bool readIPv4SourceDest(const unsigned char* packet,
+                                       const uint32_t caplen,
+                                       size_t* offset,
+                                       LocalRemoteAddrs& addrs) {
+     const size_t ipv4_len =  (((struct ip *)packet)->ip_hl << 2);
+     if (caplen < (*offset + ipv4_len)) {
+         TLOG << "Invalid packet (IPv4 header). Packet length: " << caplen
+         << ". Offset: " << *offset;
+         return false;
+     }
+     
+     const struct ip* ip = reinterpret_cast<const struct ip*>(packet + *offset);
+     *offset += ipv4_len;
+     
+     char local[INET_ADDRSTRLEN], remote[INET_ADDRSTRLEN];
+     inet_ntop(AF_INET, &ip->ip_dst, local, INET_ADDRSTRLEN);
+     inet_ntop(AF_INET, &ip->ip_src, remote, INET_ADDRSTRLEN);
+     addrs.first = std::string(local);
+     addrs.second = std::string(remote);
+     
+     return true;
+ }
+
+ static inline bool readIPv6SourceDest(const unsigned char* packet,
+                                       const uint32_t caplen,
+                                       size_t* offset,
+                                       LocalRemoteAddrs& addrs) {
+     if (caplen < (*offset + kIPv6Length)) {
+         TLOG << "Invalid packet (IPv6 header). Packet length: " << caplen
+         << ". Offset: " << *offset;
+         return false;
+     }
+     
+     const struct ip6_hdr* ip =
+     reinterpret_cast<const struct ip6_hdr*>(packet + *offset);
+     *offset += kIPv6Length;
+     
+     char local[INET6_ADDRSTRLEN], remote[INET6_ADDRSTRLEN];
+     inet_ntop(AF_INET6, &ip->ip6_dst, local, INET6_ADDRSTRLEN);
+     inet_ntop(AF_INET6, &ip->ip6_src, remote, INET6_ADDRSTRLEN);
+     addrs.first = std::string(local);
+     addrs.second = std::string(remote);
+     
+     return true;
+ }
+
+ 
+ void HTTPLookupEventPublisher::processPacket(unsigned char* args,
```

Comment:
There is quite a bit of logic in this function. It should be split into multiple class static methods so they can be better tested and fuzzed. Since it looks like it can fail in several ways, it should return Status objects